components/camel-xj/src/main/java/org/apache/camel/component/xj/XmlSourceHandlerFactoryImpl.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.xj;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.ExpectedBodyTypeException;
+import org.apache.camel.RuntimeTransformException;
+import org.apache.camel.TypeConverter;
+import org.apache.camel.support.builder.xml.XMLConverterHelper;
+import org.w3c.dom.Node;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Source;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.sax.SAXSource;
+import javax.xml.transform.stax.StAXSource;
+import javax.xml.transform.stream.StreamSource;
+import java.io.InputStream;
+
+public class XmlSourceHandlerFactoryImpl implements SourceHandlerFactory {
+
+    private XMLConverterHelper converter = new XMLConverterHelper();
+    private boolean isFailOnNullBody = true;
+
+    public void setFailOnNullBody(boolean failOnNullBody) {
+        isFailOnNullBody = failOnNullBody;
+    }
+
+    public boolean isFailOnNullBody() {
+        return isFailOnNullBody;
+    }
+
+    @Override
+    public Source getSource(Exchange exchange) throws Exception {
+        // only convert to input stream if really needed
+        if (isInputStreamNeeded(exchange)) {
+            InputStream is = exchange.getIn().getBody(InputStream.class);
+            return getSource(exchange, is);
+        } else {
+            Object body = exchange.getIn().getBody();
+            return getSource(exchange, body);
+        }
+    }
+
+    /**
+     * Checks whether we need an {@link InputStream} to access the message body.
+     * <p/>
+     * Depending on the content in the message body, we may not need to convert
+     * to {@link InputStream}.
+     *
+     * @param exchange the current exchange
+     * @return <tt>true</tt> to convert to {@link InputStream} beforehand converting to {@link Source} afterwards.
+     */
+    protected boolean isInputStreamNeeded(Exchange exchange) {
+        Object body = exchange.getIn().getBody();
```

Comment:
Lines 70-90 and 450-470 are copy-pasted. Is it possible to refactor this to method?

Reply:
obsolete.